### PR TITLE
Consistent support for Firebase Cloud Messaging

### PR DESCRIPTION
**Pull request by Antwan86**
This PR adds improved support for Firebase Cloud Messaging, sending the proper messages payload. Addresses #349 #343 amongst others.

Summary
---------

Doc is updated to reflect the changes, but here is the summary :

- Any FCM device (ie with `cloud_message_type='FCM'`) will now send appropriate payload using [notification/data formated messages](https://firebase.google.com/docs/cloud-messaging/concept-options#notifications_and_data_messages).
- Valid notification parameters are automatically retrieved and send in the `notification` json object, and can be defined in two different manners :
   - Using the `extra` dictionnary : `d.send_message("msg", extra={"title":"something"})` 
   - Using directly keyword args : `d.send_message("msg", title="something")` 
- The other parameters are sent in `data` json object.
- GCM devices will keep send the former format (using `data` json object only). This can also be enforced with FCM devices passing the `use_fcm_notifications=False` options.

Cleanup / Design choices
-------------------------

1.  Readme is updated, so FCM is now mentioned everywhere and referred as the preferred method to use Google Cloud. This is however not defaulted when creating new devices, so it won't break any existing installation.
2. The `gcm.py` connector is now using exclusively `application/json` API calls, and passing a list of `registration_ids` as a target. As a consequence the unused plaintext calls using `application/x-www-form-urlencoded` are removed, as well as the unnecessary duplicate code
3. The `send_message` from `gcm.py` can now expect either a list of `registration_id`, or a single `registration_id`, and will transparently use the proper one. As a consequence this removes the needs of a duplicated `send_bulk_message` method. (For compatibility purposes it is still declared as an alias).
4. Some internal methods have been renamed (removed `gcm` or `json` mentions) to reflect their agnosticism.

Happy to discuss these choices of course, but I think they're all minor and bring cleanliness.

### CODE REVIEW

**Comment by jleclanche** [comment]
Fantastic PR. Had a cursory glance, code looks good, but I'd like a second pair of eyes on it. @jamaalscarlett can you take a look as well?

I'm happy to merge it but I'd like to ask you to split it into distinctive commits, per CONTRIBUTING.md. If any of these (fairly consequent) changes end up being problematic, being able to easily revert them is a big deal.

Could you also add an "unreleased" entry in CHANGELOG.rst detailing the user-facing changes?

Cheers.

**Comment by Antwan86** [comment]
I've updated the changelog. Happy to spend more time on reverse-splitting the commits as it [was before squashing](https://github.com/buzzhire/django-push-notifications/commits/feature/fcm-improvements), but it will be quite time consuming to rewrite valid tests for all of them, so I'll do it only if it's really needed...

This whole PR is just about 1 consistent change : parsing the options differently to generate a different payload. The rest is just cleaning/renaming/changing doc.

**Comment by jleclanche** [comment]
I'm more concerned about splitting the 4 points you mentioned in your cleanup/design choices paragraph. Don't worry about the tests too much as long as they pass at the end of it.

**Comment by jamaalscarlett** [comment]
It looks good.  I'll take a close look when I get home tonight.

File: CHANGELOG.rst [hunk 1]
```diff
@@ -1,3 +1,8 @@
+Unreleased
```
Comment (jamaalscarlett) [inline CHANGELOG.rst]: @jleclanche Do we need the changelog entry?  It looks like they are reserved for releases.

File: README.rst [hunk 2]
```diff
@@ -3,18 +3,18 @@ django-push-notifications
...
-A minimal Django app that implements Device models that can send messages through APNS, GCM and WNS.
+A minimal Django app that implements Device models that can send messages through APNS, FCM/GCM and WNS.
 
 The app implements three models: ``GCMDevice``, ``APNSDevice`` and ``WNSDevice``. Those models share the same attributes:
  - ``name`` (optional): A name for the device.
  - ``active`` (default True): A boolean that determines whether the device will be sent notifications.
  - ``user`` (optional): A foreign key to auth.User, if you wish to link the device to a specific user.
  - ``device_id`` (optional): A UUID for the device obtained from Android/iOS/Windows APIs, if you wish to uniquely identify it.
- - ``registration_id`` (required): The GCM registration id or the APNS token for the device.
+ - ``registration_id`` (required): The FCM registration id or the APNS token for the device.
```
Comment (jamaalscarlett) [inline README.rst]: There is so formatting incongruity here.  On line 6 you use FCM/GCM, while on this and subsequent lines you use only FCM.  Should probably keep it FCM/GCM.

File: README.rst [hunk 3]
```diff
@@ -65,28 +66,35 @@ Settings list
 -------------
 All settings are contained in a ``PUSH_NOTIFICATIONS_SETTINGS`` dict.
 
-In order to use GCM, you are required to include ``GCM_API_KEY``.
+In order to use FCM, you are required to include ``FCM_API_KEY``.
```
Comment (jamaalscarlett) [inline README.rst]: See comment above.  While GCM is deprecated, it does still work, best to leave those instructions alone and add to them.

File: README.rst [hunk 4]
```diff
@@ -135,20 +143,61 @@ value per user. Assuming User model has a method get_badge returning badge count
...
+
+	# Send a notification message with options
+	fcm_device.send_message("This is a message", time_to_live=3600)
+
+	# Send a data message only
+	fcm_device.send_message(None, extra={"other": "content", "misc": "data"})
+
+You can disable this default behaviour by setting ``use_fcm_notifications`` to ``False``.
+
+.. code-block:: python
```
Comment (jamaalscarlett) [inline README.rst]: typo?

**Comment by jamaalscarlett** [comment]
Other than the nit picks about the comments, this looks good to merge.

**Comment by jamaalscarlett** [comment]
@Antwan86 bump

**Comment by Antwan86** [comment]
Hey ! Thanks for the review and sorry for the delay, I updated the readme, will split the commits later, hopefully this weekend.

**Comment by jleclanche** [comment]
Is this ready to merge?

**Comment by Antwan86** [comment]
I replaced this PR by #360 which includes a better commit history.
I think it's ready to merge now unless you see other issues.

Also just FYI, it doesn't replace a proper review and tests but I've been using this code in prod for a month without issue.

**Pull request closed without merging**